### PR TITLE
Take geographic distance into account for server list ordering

### PR DESCRIPTION
**Pull request by sfan5**
Change the sorting to take the estimated latency into account. The original sorting of the server list is still honored.
This makes more sense to give players a better experience so they can play on low latency servers. The ping reported by the serverlist is not usable for this since it's measured from its own location (Netherlands).

geolocation is provided by the serverlist (minetest/serverlist#34, already live)

normalized ranking according to https://github.com/minetest/minetest/pull/8688#issuecomment-521528999 (@sofar)

<b>Example</b>: how the server list would look for North American players right now
(continent in angle brackets, red numbers is rank on global list)
grafik

## To do

This PR is Ready for Review.

- [ ] maybe tune the weights
- [x] store retrieved geoip outside of mainmenu, since the Lua state is recreated when you re-enter the menu

## How to test

To test this properly you'll want to do this ↓↓ and open https://servers.minetest.net/ in a new tab to compare.
```diff
diff --git a/builtin/mainmenu/common.lua b/builtin/mainmenu/common.lua
--- a/builtin/mainmenu/common.lua
+++ b/builtin/mainmenu/common.lua
@@ -75,6 +75,9 @@ function render_serverlist_row(spec)
                        text = text .. ":" .. spec.port
                end
        end
+       if spec.geo_continent then
+               text = "<" .. spec.geo_continent .. "> " .. text
+       end
 
        local grey_out = not spec.is_compatible
 
diff --git a/builtin/mainmenu/serverlistmgr.lua b/builtin/mainmenu/serverlistmgr.lua
--- a/builtin/mainmenu/serverlistmgr.lua
+++ b/builtin/mainmenu/serverlistmgr.lua
@@ -17,7 +17,7 @@
 
 serverlistmgr = {
        -- continent code we detected for ourselves
-       my_continent = core.get_once("continent"),
+       my_continent = core.get_once("continent") or "NA", -- enter your favorite continent here
 
        -- list of locally favorites servers
        favorites = nil,

```

### CODE REVIEW

**Comment by tigercoding56** [comment]
perhabs it could also show the latency next to the servers name  if a option in minetest.conf is enabled

**Comment by SmallJoker** [comment]
@tigercoding56 This latency information is not accurate on a per-server basis but helps to roughly sort the list. I do not think that it's something that should be displayed.

File: builtin/mainmenu/serverlistmgr.lua [hunk 1]
```diff
@@ -56,6 +129,35 @@ function serverlistmgr.sync()
 		return
 	end
 
+	-- only fetched once per session
+	if not serverlistmgr.my_continent and not geoip_downloading then
+		geoip_downloading = true
+		core.handle_async(
```
Comment (SmallJoker) [inline builtin/mainmenu/serverlistmgr.lua:135]: What's the advantage of wrapping handle_async around, as opposed to the async HTTP API?

Comment (sfan5) [reply builtin/mainmenu/serverlistmgr.lua:135]: The mainmenu does not have the async api, that's why

File: builtin/mainmenu/serverlistmgr.lua [hunk 2]
```diff
@@ -15,28 +15,101 @@
...
+local WEIGHT_SORT = 2
+-- how much the estimated latency contributes to the final ranking
+local WEIGHT_LATENCY = 1
+
 local function order_server_list(list)
-	local res = {}
-	--orders the favorite list after support
-	for i = 1, #list do
-		local fav = list[i]
-		if is_server_protocol_compat(fav.proto_min, fav.proto_max) then
```
[the remark this answers is not in the document]
Comment (sfan5) [reply builtin/mainmenu/serverlistmgr.lua:26]: yes, the returned list is split into favorite/compatible/incompatible by later code

File: builtin/mainmenu/common.lua [hunk 3]
```diff
@@ -45,6 +45,26 @@ local function configure_selected_world_params(idx)
...
+}
+function estimate_continent_latency(own, spec)
+	local there = spec.geo_continent
+	if not own or not there then
+		return nil
+	end
+	if not own or not there or own == there then
+		return 0
+	end
+	return own > there and latency_matrix[there][own] or latency_matrix[own][there]
```
[the remark this answers is not in the document]
Comment (sfan5) [reply builtin/mainmenu/common.lua:65]: I really don't expect this to be an issue ever but I'll change it

**Review by TurkeyMcMac — APPROVED** [review]
Looks good. The weights seem fine, though I don't know how ideal weights would be determined,

**Pull request merged**